<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Connecting Programs and Compressing output | A minimal quantitative RNAseq Pipeline</title>
  <meta name="description" content="A minimal guide to performing quantitative RNAseq" />
  <meta name="generator" content="bookdown 0.13 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Connecting Programs and Compressing output | A minimal quantitative RNAseq Pipeline" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A minimal guide to performing quantitative RNAseq" />
  <meta name="github-repo" content="TeamMacLean/minimal_quant_rna_seq" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Connecting Programs and Compressing output | A minimal quantitative RNAseq Pipeline" />
  
  <meta name="twitter:description" content="A minimal guide to performing quantitative RNAseq" />
  

<meta name="author" content="Dan MacLean" />


<meta name="date" content="2020-01-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="filtering-badly-aligned-reads.html"/>
<link rel="next" href="automating-the-process.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A minimal alignment guide</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About this course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> R and RStudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="index.html"><a href="index.html#sample-reference-genome-and-reads"><i class="fa fa-check"></i><b>1.1.2</b> Sample reference genome and reads</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Counting Aligned Reads in Genomic Regions</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#about-this-chapter"><i class="fa fa-check"></i><b>2.1</b> About this chapter</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#counting-the-number-of-reads-that-have-aligned-to-gene-regions"><i class="fa fa-check"></i><b>2.2</b> Counting the number of reads that have aligned to gene regions</a><ul>
<li class="chapter" data-level="2.2.1" data-path="intro.html"><a href="intro.html#atacr"><i class="fa fa-check"></i><b>2.2.1</b> atacR</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#preparing-the-input"><i class="fa fa-check"></i><b>2.3</b> Preparing the input</a><ul>
<li class="chapter" data-level="2.3.1" data-path="intro.html"><a href="intro.html#the-gff-file"><i class="fa fa-check"></i><b>2.3.1</b> The GFF file</a></li>
<li class="chapter" data-level="2.3.2" data-path="intro.html"><a href="intro.html#the-sample-information-file"><i class="fa fa-check"></i><b>2.3.2</b> The Sample Information file</a></li>
<li class="chapter" data-level="2.3.3" data-path="intro.html"><a href="intro.html#the-bam-files"><i class="fa fa-check"></i><b>2.3.3</b> The BAM files</a></li>
<li class="chapter" data-level="2.3.4" data-path="intro.html"><a href="intro.html#sample-files-for-this-chapter"><i class="fa fa-check"></i><b>2.3.4</b> Sample files for this chapter</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#running-make_counts"><i class="fa fa-check"></i><b>2.4</b> Running <code>make_counts()</code></a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#summaries-and-diagnostic-plots"><i class="fa fa-check"></i><b>2.5</b> Summaries and Diagnostic plots</a><ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#gene-count-plots"><i class="fa fa-check"></i><b>2.5.1</b> Gene Count Plots</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#comparing-samples-with-pca"><i class="fa fa-check"></i><b>2.5.2</b> Comparing Samples with PCA</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#extracting-and-saving-the-count-matrix"><i class="fa fa-check"></i><b>2.6</b> Extracting and saving the count matrix</a><ul>
<li class="chapter" data-level="2.6.1" data-path="intro.html"><a href="intro.html#saving-to-an-r-rds-file"><i class="fa fa-check"></i><b>2.6.1</b> Saving to an R RDS file</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="running-minimap2.html"><a href="running-minimap2.html"><i class="fa fa-check"></i><b>3</b> Running <code>minimap2</code></a><ul>
<li class="chapter" data-level="3.1" data-path="running-minimap2.html"><a href="running-minimap2.html#the-minimap2-command-and-options"><i class="fa fa-check"></i><b>3.1</b> The <code>minimap2</code> command and options</a></li>
<li class="chapter" data-level="3.2" data-path="running-minimap2.html"><a href="running-minimap2.html#further-reading"><i class="fa fa-check"></i><b>3.2</b> Further Reading</a><ul>
<li class="chapter" data-level="3.2.1" data-path="running-minimap2.html"><a href="running-minimap2.html#the-operator"><i class="fa fa-check"></i><b>3.2.1</b> The <code>&gt;</code> operator</a></li>
<li class="chapter" data-level="3.2.2" data-path="running-minimap2.html"><a href="running-minimap2.html#minimap2-further-instructions-and-github"><i class="fa fa-check"></i><b>3.2.2</b> <code>minimap2</code> further instructions and github</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html"><i class="fa fa-check"></i><b>4</b> Filtering Badly Aligned Reads</a><ul>
<li class="chapter" data-level="4.1" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#sam-format"><i class="fa fa-check"></i><b>4.1</b> SAM Format</a></li>
<li class="chapter" data-level="4.2" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#samtools"><i class="fa fa-check"></i><b>4.2</b> <code>samtools</code></a></li>
<li class="chapter" data-level="4.3" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#the-samtools-command-and-options"><i class="fa fa-check"></i><b>4.3</b> The <code>samtools</code> command and options</a></li>
<li class="chapter" data-level="4.4" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#checking-the-filtering"><i class="fa fa-check"></i><b>4.4</b> Checking the filtering</a></li>
<li class="chapter" data-level="4.5" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#are-we-done"><i class="fa fa-check"></i><b>4.5</b> Are we done?</a></li>
<li class="chapter" data-level="4.6" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#further-reading-1"><i class="fa fa-check"></i><b>4.6</b> Further Reading</a><ul>
<li class="chapter" data-level="4.6.1" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#sam-format-1"><i class="fa fa-check"></i><b>4.6.1</b> SAM Format</a></li>
<li class="chapter" data-level="4.6.2" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#mapping-quality"><i class="fa fa-check"></i><b>4.6.2</b> Mapping Quality</a></li>
<li class="chapter" data-level="4.6.3" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#flags"><i class="fa fa-check"></i><b>4.6.3</b> Flags</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html"><i class="fa fa-check"></i><b>5</b> Connecting Programs and Compressing output</a><ul>
<li class="chapter" data-level="5.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#bam-files"><i class="fa fa-check"></i><b>5.1</b> BAM Files</a></li>
<li class="chapter" data-level="5.2" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#connecting-program-input-and-output-with-pipes"><i class="fa fa-check"></i><b>5.2</b> Connecting Program Input and Output With Pipes</a></li>
<li class="chapter" data-level="5.3" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#from-reads-to-filtered-alignments-in-one-step"><i class="fa fa-check"></i><b>5.3</b> From reads to filtered alignments in one step</a><ul>
<li class="chapter" data-level="5.3.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#from-reads-to-filtered-alignments-in-a-bam-file-in-one-step"><i class="fa fa-check"></i><b>5.3.1</b> From reads to filtered alignments in a BAM file in one step</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#sorting-bam-files"><i class="fa fa-check"></i><b>5.4</b> Sorting BAM files</a><ul>
<li class="chapter" data-level="5.4.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#automatically-deleting-the-unsorted-bam"><i class="fa fa-check"></i><b>5.4.1</b> Automatically deleting the unsorted BAM</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#indexing-the-sorted-bam"><i class="fa fa-check"></i><b>5.5</b> Indexing the sorted BAM</a></li>
<li class="chapter" data-level="5.6" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#further-reading-2"><i class="fa fa-check"></i><b>5.6</b> Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="automating-the-process.html"><a href="automating-the-process.html"><i class="fa fa-check"></i><b>6</b> Automating The Process</a><ul>
<li class="chapter" data-level="6.1" data-path="automating-the-process.html"><a href="automating-the-process.html#shell-scripts"><i class="fa fa-check"></i><b>6.1</b> Shell scripts</a><ul>
<li class="chapter" data-level="6.1.1" data-path="automating-the-process.html"><a href="automating-the-process.html#using-nano-to-create-a-shell-script"><i class="fa fa-check"></i><b>6.1.1</b> Using <code>nano</code> to create a shell script</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="automating-the-process.html"><a href="automating-the-process.html#creating-a-script-that-automates-our-alignment-pipeline."><i class="fa fa-check"></i><b>6.2</b> Creating a script that automates our alignment pipeline.</a></li>
<li class="chapter" data-level="6.3" data-path="automating-the-process.html"><a href="automating-the-process.html#running-the-script"><i class="fa fa-check"></i><b>6.3</b> Running the script</a></li>
<li class="chapter" data-level="6.4" data-path="automating-the-process.html"><a href="automating-the-process.html#running-on-different-input-files"><i class="fa fa-check"></i><b>6.4</b> Running on different input files</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html"><i class="fa fa-check"></i><b>7</b> Running an alignment on the HPC</a><ul>
<li class="chapter" data-level="7.1" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#an-hpc-is-a-group-of-slave-computers-under-control-of-a-master-computer"><i class="fa fa-check"></i><b>7.1</b> An HPC is a group of slave computers under control of a master computer</a></li>
<li class="chapter" data-level="7.2" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#logging-into-the-submission-node"><i class="fa fa-check"></i><b>7.2</b> Logging into the submission node</a></li>
<li class="chapter" data-level="7.3" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#preparing-a-job"><i class="fa fa-check"></i><b>7.3</b> Preparing a job</a><ul>
<li class="chapter" data-level="7.3.1" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-sbatch-options"><i class="fa fa-check"></i><b>7.3.1</b> The <code>#SBATCH</code> options</a></li>
<li class="chapter" data-level="7.3.2" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-source-options"><i class="fa fa-check"></i><b>7.3.2</b> The <code>source</code> options</a></li>
<li class="chapter" data-level="7.3.3" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-srun-command"><i class="fa fa-check"></i><b>7.3.3</b> The <code>srun</code> command</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#submitting-with-sbatch"><i class="fa fa-check"></i><b>7.4</b> Submitting with <code>sbatch</code></a></li>
<li class="chapter" data-level="7.5" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#checkout-tasks"><i class="fa fa-check"></i><b>7.5</b> Checkout tasks</a></li>
<li class="chapter" data-level="7.6" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#further-reading-3"><i class="fa fa-check"></i><b>7.6</b> Further Reading</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A minimal quantitative RNAseq Pipeline</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="connecting-programs-and-compressing-output" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Connecting Programs and Compressing output</h1>
<p>Now that we’ve been through the whole alignment and filtering pipeline, let’s look at the output. Specifically lets compare the sizes of the files we used. Recall that we can do that with <code>ls -alh</code></p>
<p>On my folder I get this (some columns and files removed for clarity)</p>
<pre><code>49M 29 Nov 10:46 aln.filtered.sam
59M 28 Nov 16:28 aln.sam
5.0M  2 Jul 15:04 ecoli_genome.fa
18M 28 Nov 15:53 ecoli_left_R1.fq
18M 28 Nov 15:53 ecoli_right_R2.fq</code></pre>
<p>The file sizes are in the left-most column. Check out the relative size of the two read files (18M each) and the alignment SAM files (59M and 49M). The output file is much larger than the input. This has implications for storage when the files are really large (many GB) and there are lots of them. The disk space gets used really quickly. Consider also the redundancy we have - that <code>aln.filtered.sam</code> is the one we’re interested in, not the <code>aln.sam</code> so it is taking up unnecesary disk space. It’s easy to see that when you are doing a real experiment with lots of samples and hundreds of GB file size, you’re going to eat up disk space. Also larger files take longer to process, so you’re going to have a long wait. This has implications too when you get to later stages in the analysis</p>
<p>In this chapter we’re going to look at a technique for reducing those housekeeping overheads and speeding things up.</p>
<div id="bam-files" class="section level2">
<h2><span class="header-section-number">5.1</span> BAM Files</h2>
<p>BAM files are a binary compressed version of SAM files. They contain identical information in a more computer friendly way. This means that people can’t read it, but it is rare in practice that you’ll directly read much of a SAM file with your own eyes. Let’s look at the command to do that</p>
<pre><code>samtools view -S -b aln.filtered.sam &gt; aln.filtered.bam
</code></pre>
<p>Again we’re using <code>samtools view</code> and our options are <code>-S</code> which means SAM format input and the new one is <code>-b</code> means BAM format output. Our input file is <code>aln.filtered</code> and we’re sending the output to <code>aln.filtered.bam</code>.</p>
<p>If we check the files with <code>ls -alh</code> now we get</p>
<pre><code>9.2M 29 Nov 14:05 aln.filtered.bam
49M 29 Nov 10:46 aln.filtered.sam
59M 28 Nov 16:28 aln.sam
5.0M  2 Jul 15:04 ecoli_genome.fa
18M 28 Nov 15:53 ecoli_left_R1.fq
18M 28 Nov 15:53 ecoli_right_R2.fq</code></pre>
<p>The BAM file is about a fifth of the size of the SAM file. So we can save space in this way. We have another trick up our sleeve though. We can connect together command lines, so that we don’t have to create intermediate files - this reduces the number of files we have to save. We can do this by using something called pipes.</p>
</div>
<div id="connecting-program-input-and-output-with-pipes" class="section level2">
<h2><span class="header-section-number">5.2</span> Connecting Program Input and Output With Pipes</h2>
<p>Most command line programs print their results straight out without sending it to a file. This seems strange, but it adds a lot of flexibility. If also set up our programs to read in this output then we can connect them together. We can do this with pipes. The usual way to do this is to use the <code>|</code> operator. Let’s look at a common example.</p>
<p>Here we’ll use the command <code>ls</code> and <code>shuf</code> to see how this works. We know <code>ls</code> will ‘list’ our directory contents, <code>shuf</code> shuffles lines of text sent to it. If we use <code>|</code> in between we can connect the output of one to the other. Try running <code>ls</code> a couple of times to verify you get the same output both times and then try this
a few times</p>
<pre><code>ls | shuf</code></pre>
<p>you should get different output everytime. The important thing to note is that <code>shuf</code> is doing its job on the data sent from <code>ls</code>, which sends consistent data every time. We don’t have to create an intermediate file for <code>shuf</code> to work from. The <code>|</code> character joing two commands is the key.</p>
<p>We can apply this to our <code>minimap2</code> and <code>samtools</code> commands.</p>
</div>
<div id="from-reads-to-filtered-alignments-in-one-step" class="section level2">
<h2><span class="header-section-number">5.3</span> From reads to filtered alignments in one step</h2>
<p>So let’s try reducing the original alignment pipeline to one step with pipes. We’ll work in the BAM file bit later.</p>
<p>Simply take away the output file names (except the last one!) and replace with pipes. It looks like this</p>
<pre><code>minimap2 -ax sr ecoli_genome.fa ecoli_left_R1.fq ecoli_right_R2.fq | samtools view -S -h -q 25 -f 3 &gt; aln.filtered.from_pipes.sam</code></pre>
<p>when you do <code>ls -alh</code> you should see the new <code>aln.filtered.from_pipes.sam</code> file, its size is identical to the file we generated when we created the intermediate <code>aln.sam</code> file, but this time we didnt need to, saving that disk space.</p>
<div id="from-reads-to-filtered-alignments-in-a-bam-file-in-one-step" class="section level3">
<h3><span class="header-section-number">5.3.1</span> From reads to filtered alignments in a BAM file in one step</h3>
<p>Let’s modify the command to give us BAM not SAM, saving a further step. We already know that <code>samtools view</code> can output BAM instead of SAM, so lets add that option (<code>-b</code>) in to the <code>samtools</code> part.</p>
<pre><code>minimap2 -ax sr ecoli_genome.fa ecoli_left_R1.fq ecoli_right_R2.fq | samtools view -S -h -b -q 25 -f 3 &gt; aln.filtered.from_pipes.bam</code></pre>
<p>If you check the files with <code>ls -alh</code> now you should see that you have the new <code>aln.filtered.from_pipes.bam</code> file with no extra intermediate file and the smallest possible output file. Congratulations, you know now the fastest and most optimal way to make alignments and filter them.</p>
</div>
</div>
<div id="sorting-bam-files" class="section level2">
<h2><span class="header-section-number">5.4</span> Sorting BAM files</h2>
<p>In practice a BAM file of alignments needs to be ordered with the alignments at the start of the first chromosome at the start of the file and the alignments on the end of the last chromosome at the end of the file. This is for computational reasons we don’t need to worry about, but it does mean we need to do another sorting step to make our files useful downstream.</p>
<p>Because all the alignments need to be present before we can start we can’t use the pipe technique above. So we use an input and output file. The command is <code>samtools sort</code> and looks like this.</p>
<pre><code>samtools sort aln.filtered.from_pipes.bam -o aln.filtered.from_pipes.sorted.bam</code></pre>
<p>Doing <code>ls -alh</code> shows a new sorted BAM <code>aln.filtered.from_pipes.sorted.bam</code> that contains the same information but is actually a little smaller due to being sorted. We can safely delete the unsorted version of the BAM file.</p>
<div id="automatically-deleting-the-unsorted-bam" class="section level3">
<h3><span class="header-section-number">5.4.1</span> Automatically deleting the unsorted BAM</h3>
<p>If the sorting goes fine, we have two BAM files with essentially the same information and don’t need the unsorted file. We can of course remove this with <code>rm aln.filtered.from_pipes</code>. A neat space saving trick is to combine the <code>rm</code> step with the successful completion of the sort. We can do this by joining the commands with <code>&amp;&amp;</code>.</p>
<p>That looks like this</p>
<pre><code>samtools sort aln.filtered.from_pipes.bam -o aln.filtered.from_pipes.sorted.bam &amp;&amp; rm aln.filtered.from_pipes.bam</code></pre>
<p>The <code>&amp;&amp;</code> doesn’t connect the data between the two commands, it just doesn’t let the second one start until the first one finishes successfully (computers have an internal concept of whether a command finished properly).</p>
<p>This means if the <code>samtools sort</code> goes wrong the <code>rm</code> part will not run and the input file won’t be deleted so you won’t have to remake it. This is especially useful later when we wrap all this into an automatic script.</p>
</div>
</div>
<div id="indexing-the-sorted-bam" class="section level2">
<h2><span class="header-section-number">5.5</span> Indexing the sorted BAM</h2>
<p>Many downstream applications need the BAM file to have an index, so they can quickly jump to a particular part of the reference chromosome. This is a tiny file and we usually don’t need to worry about it. To generate it use <code>samtools index</code></p>
<pre><code>samtools index aln.filtered.from_pipes.sorted.bam</code></pre>
<p>Using <code>ls -lah</code> we can see a tiny file called <code>aln.filtered.from_pipes.sorted.bam.bai</code>, this is the index.</p>
</div>
<div id="further-reading-2" class="section level2">
<h2><span class="header-section-number">5.6</span> Further Reading</h2>
<p>For a primer on some more aspects of <code>samtools</code> see this <a href="http://quinlanlab.org/tutorials/samtools/samtools.html">tutorial</a></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="filtering-badly-aligned-reads.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="automating-the-process.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["Basic Read Alignment.pdf", "Basic Read Alignment.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
