<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>A minimal quantitative RNAseq pipeline</title>
  <meta name="description" content="A minimal guide to performing quantitative RNAseq" />
  <meta name="generator" content="bookdown 0.13 and GitBook 2.6.7" />

  <meta property="og:title" content="A minimal quantitative RNAseq pipeline" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A minimal guide to performing quantitative RNAseq" />
  <meta name="github-repo" content="TeamMacLean/minimal_quant_rna_seq" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="A minimal quantitative RNAseq pipeline" />
  
  <meta name="twitter:description" content="A minimal guide to performing quantitative RNAseq" />
  

<meta name="author" content="Dan MacLean" />


<meta name="date" content="2020-01-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="intro.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A minimal alignment guide</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About this course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> R and RStudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="index.html"><a href="index.html#sample-bam-file-and-counts-files"><i class="fa fa-check"></i><b>1.1.2</b> Sample BAM file and counts files</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Counting Aligned Reads in Genomic Regions</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#about-this-chapter"><i class="fa fa-check"></i><b>2.1</b> About this chapter</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#counting-the-number-of-reads-that-have-aligned-to-gene-regions"><i class="fa fa-check"></i><b>2.2</b> Counting the number of reads that have aligned to gene regions</a><ul>
<li class="chapter" data-level="2.2.1" data-path="intro.html"><a href="intro.html#atacr"><i class="fa fa-check"></i><b>2.2.1</b> atacR</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#preparing-the-input"><i class="fa fa-check"></i><b>2.3</b> Preparing the input</a><ul>
<li class="chapter" data-level="2.3.1" data-path="intro.html"><a href="intro.html#the-gff-file"><i class="fa fa-check"></i><b>2.3.1</b> The GFF file</a></li>
<li class="chapter" data-level="2.3.2" data-path="intro.html"><a href="intro.html#the-sample-information-file"><i class="fa fa-check"></i><b>2.3.2</b> The Sample Information file</a></li>
<li class="chapter" data-level="2.3.3" data-path="intro.html"><a href="intro.html#the-bam-files"><i class="fa fa-check"></i><b>2.3.3</b> The BAM files</a></li>
<li class="chapter" data-level="2.3.4" data-path="intro.html"><a href="intro.html#sample-files-for-this-chapter"><i class="fa fa-check"></i><b>2.3.4</b> Sample files for this chapter</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#running-make_counts"><i class="fa fa-check"></i><b>2.4</b> Running <code>make_counts()</code></a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#summaries-and-diagnostic-plots"><i class="fa fa-check"></i><b>2.5</b> Summaries and Diagnostic plots</a><ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#gene-count-plots"><i class="fa fa-check"></i><b>2.5.1</b> Gene Count Plots</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#comparing-samples-with-pca"><i class="fa fa-check"></i><b>2.5.2</b> Comparing Samples with PCA</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#extracting-and-saving-the-count-matrix"><i class="fa fa-check"></i><b>2.6</b> Extracting and saving the count matrix</a><ul>
<li class="chapter" data-level="2.6.1" data-path="intro.html"><a href="intro.html#saving-to-an-r-rds-file"><i class="fa fa-check"></i><b>2.6.1</b> Saving to an R RDS file</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="running-deseq2.html"><a href="running-deseq2.html"><i class="fa fa-check"></i><b>3</b> Running <code>DESeq2</code></a><ul>
<li class="chapter" data-level="3.1" data-path="running-deseq2.html"><a href="running-deseq2.html#about-this-chapter-1"><i class="fa fa-check"></i><b>3.1</b> About this chapter</a></li>
<li class="chapter" data-level="3.2" data-path="running-deseq2.html"><a href="running-deseq2.html#getting-the-count-matrix-and-describing-the-experiment-for-deseq2"><i class="fa fa-check"></i><b>3.2</b> Getting the count matrix and describing the experiment for DESeq2</a><ul>
<li class="chapter" data-level="3.2.1" data-path="running-deseq2.html"><a href="running-deseq2.html#the-count-matrix"><i class="fa fa-check"></i><b>3.2.1</b> The count matrix</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="running-deseq2.html"><a href="running-deseq2.html#the-grouping-object"><i class="fa fa-check"></i><b>3.3</b> The ‘grouping’ object</a></li>
<li class="chapter" data-level="3.4" data-path="running-deseq2.html"><a href="running-deseq2.html#running-deseq2-1"><i class="fa fa-check"></i><b>3.4</b> Running DESeq2</a></li>
<li class="chapter" data-level="3.5" data-path="running-deseq2.html"><a href="running-deseq2.html#saving-the-results"><i class="fa fa-check"></i><b>3.5</b> Saving the results</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="next-steps.html"><a href="next-steps.html"><i class="fa fa-check"></i><b>4</b> Next Steps</a><ul>
<li class="chapter" data-level="4.1" data-path="next-steps.html"><a href="next-steps.html#about-this-chapter-2"><i class="fa fa-check"></i><b>4.1</b> About this chapter</a></li>
<li class="chapter" data-level="4.2" data-path="next-steps.html"><a href="next-steps.html#the-results-data-frame"><i class="fa fa-check"></i><b>4.2</b> The results data frame</a><ul>
<li class="chapter" data-level="4.2.1" data-path="next-steps.html"><a href="next-steps.html#which-p-value"><i class="fa fa-check"></i><b>4.2.1</b> Which <em>p</em> value?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="next-steps.html"><a href="next-steps.html#filtering-rows-with-significant-p-values"><i class="fa fa-check"></i><b>4.3</b> Filtering rows with significant <em>p</em> values</a><ul>
<li class="chapter" data-level="4.3.1" data-path="next-steps.html"><a href="next-steps.html#filtering-up-and-down-genes"><i class="fa fa-check"></i><b>4.3.1</b> Filtering UP and DOWN genes</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="next-steps.html"><a href="next-steps.html#finding-gene-annotations"><i class="fa fa-check"></i><b>4.4</b> Finding gene annotations</a><ul>
<li class="chapter" data-level="4.4.1" data-path="next-steps.html"><a href="next-steps.html#biomart"><i class="fa fa-check"></i><b>4.4.1</b> BioMart</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="next-steps.html"><a href="next-steps.html#further-questions"><i class="fa fa-check"></i><b>4.5</b> Further Questions</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A minimal quantitative RNAseq pipeline</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">A minimal quantitative RNAseq pipeline</h1>
<p class="author"><em>Dan MacLean</em></p>
<p class="date"><em>2020-01-30</em></p>
</div>
<div id="about-this-course" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> About this course</h1>
<p>In this short course we’ll look at a method for getting quantitative estimates of gene expression from RNAseq data. The course assumes that you will already have performed a read alignment so is <em>not</em> a ‘read to results’ course. The course is very brief and will show you how to use a perform a common pipeline centered around <code>DESeq</code> in <code>R</code> and <code>RStudio</code></p>
<p>I acknowledge that there are lots of other programs and methods - this course is <em>not</em> meant to be comprehensive, it is meant to get you being productive. Seek out further advice if you need to run other programs or systems. Do be encouraged though, lots of what you learn here will be applicable to other pipelines for the same job (they all run in a similar manner with similar objects) so this is a good place to start.</p>
<p>The course is intended to run on your ‘local’ machine, that is to say, your laptop or desktop computer. In general these machines will be powerful enough for most datasets though the pipeline we will learn can be easily adapted for a high performance computing environment if you need greater computational power.</p>
<div id="prerequisites" class="section level2">
<h2><span class="header-section-number">1.1</span> Prerequisites</h2>
<p>This course assumes that you are a little familiar with the basics of running R and R commands from the R console. You’ll need to know the basics of typing in commands and getting output, not much more.</p>
<div id="r-and-rstudio" class="section level3">
<h3><span class="header-section-number">1.1.1</span> R and RStudio</h3>
<div id="installing-r" class="section level4">
<h4><span class="header-section-number">1.1.1.1</span> Installing R</h4>
<p>Follow this link and install the right version for your operating system <a href="https://www.stats.bris.ac.uk/R/">https://www.stats.bris.ac.uk/R/</a></p>
</div>
<div id="installing-rstudio" class="section level4">
<h4><span class="header-section-number">1.1.1.2</span> Installing RStudio</h4>
<p>Follow this link and install the right version for your operating system <a href="https://www.rstudio.com/products/rstudio/download/">https://www.rstudio.com/products/rstudio/download/</a></p>
</div>
<div id="installing-r-packages-in-rstudio." class="section level4">
<h4><span class="header-section-number">1.1.1.3</span> Installing R packages in RStudio.</h4>
<p>You’ll need the following R packages:</p>
<ol style="list-style-type: decimal">
<li>devtools</li>
<li>atacR</li>
<li>DESeq</li>
</ol>
<p>For simplicity, install them in that order.</p>
<p>To install <code>devtools</code>:</p>
<p>Start RStudio and use the <code>Packages</code> tab in lower right panel. Click the install button (top left of the panel) and enter the package name <code>devtools</code>, then click install as in this picture</p>
<div class="figure">
<img src="fig/package_install.png" alt="Installing Packages" />
<p class="caption">Installing Packages</p>
</div>
<p>To install <code>atacR</code>:</p>
<p>Type the following into the RStudio console, <code>devtools::install_github(&quot;TeamMacLean/atacr&quot;)</code></p>
<p>To install <code>DESeq</code>:</p>
<p>Type the following into the RStudio console, <code>BiocManager::install(&quot;DESeq&quot;)</code></p>
<p>Now you are done! Everything is installed ready for you to work with. Next we need to get the sample data</p>
</div>
</div>
<div id="sample-bam-file-and-counts-files" class="section level3">
<h3><span class="header-section-number">1.1.2</span> Sample BAM file and counts files</h3>
<p>You’ll need this zip file of data: <a href="https://github.com/TeamMacLean/minimal_quantitative_rnaseq/blob/master/sample_data/sample_data.zip">sample_data.zip</a> which contains a lot of ifles including BAM files and counts. Download it, extract the files and put them into a folder on your machine. I suggest something like <code>Desktop/align_tut</code>. This will be the directory we’ll work from in the rest of the course.</p>
<p>That’s all you need to do the lesson. If you have any problems getting this going, then ask someone in the Bioinformatics Team and we’ll help.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="intro.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["Minimal Quantitative RNASeq Pipeline.pdf", "Minimal Quantitative RNASeq Pipeline.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
